chainstate/src/detail/block_invalidation/best_chain_candidates.rs

```diff
@@ -127,26 +127,28 @@ impl BestChainCandidates {
     }
 
     // Remove the block and its descendants, which must be specified in descendants_indices,
-    // from the set and add the block's parent block to the set.
+    // from the set and add the block's parent block to the set if its chain trust is not less than
+    // the specified minimum.
     #[log_error]
-    pub fn on_block_invalidated<Chs: ChainstateAccessor>(
+    pub fn remove_tree_add_parent<Chs: ChainstateAccessor>(
         &mut self,
         chs: &Chs,
-        invalidated_block_info: &Chs::BlockInfo,
+        root_block_info: &Chs::BlockInfo,
         descendant_block_infos: &[Chs::BlockInfo],
+        min_chain_trust: Uint256,
```

Comment:
Why?

Changing these algorithms needs very careful security consideration. We cannot just change them nonchalant to achieve some technical goal. Let's discuss this.

Keep in mind that adding things like this make tests reproducibility very difficult.

Reply:
This is called as a part of `find_and_activate_best_chain` call during block invalidation. Though `invalidate_block` gets called in `attempt_to_process_block` too, in that case it's only called for a stale block for which a reorg attempt has already failed, so it won't trigger another reorg. I.e. this logic will kick in only on a manual block invalidation. I don't think it can affect security.
This particular change (passing `min_chain_trust` into `on_block_invalidated` and taking it into account) is a fix for the assertion failure that we had recently when trying to manually invalidate a block (the test for this is called `test_invalidation_with_reorg_attempt_to_chain_with_lower_chain_trust`)